Adds functional tests for self service wallets.

mango-habanero opened the pull request.

mango-habanero committed:
Adds functional tests for self service wallets.

mango-habanero committed:
Merge branch 'master' into philip/self-service-wallets-func-tests

mango-habanero committed:
Merge branch 'master' into philip/self-service-wallets-func-tests

nolash commented on test/functional/app/api/test_ussd_api.py:
```diff
@@ -127,11 +141,66 @@ def test_golden_path_send_token(mocker, test_client,
+
+
+@pytest.fixture(scope='module')
+def create_temporary_user(test_client, init_database, create_organisation):
```
This is a fixture so it should be in a `conftest.py` file. Judging from context it looks like it should be in the root one.

mango-habanero replied:
I grappled with putting it there but then its only used once in this very specific context ? Should such fixtures also be in conftest ?

nolash replied:
I believe it should be in a `conftest.py` file. But that could be a separate file in this directory, too. If I recall correctly parent dirs are checked in order for that filename to include all fixtures.

nolash commented on test/functional/app/api/test_ussd_api.py:
```diff
@@ -127,11 +141,66 @@ def test_golden_path_send_token(mocker, test_client,
     assert 'END Bonyeza {} kutumia mtandao'.format(valid_service_code) in resp
+
+
+@pytest.fixture(scope='module')
```
scope function because you would be manipulating the user from scratch each test, no?

mango-habanero replied:
Yes

nolash commented on test/functional/app/api/test_ussd_api.py:
```diff
@@ -127,11 +141,66 @@ def test_golden_path_send_token(mocker, test_client,
+    g.active_organisation = organisation
+
+    # create user without a transfer account
+    user = create_user_without_transfer_account(unregistered_user_phone)
```
I see we have no test for the "attach transfer account" hack from last time. Could we have one please?

nolash commented on test/functional/app/api/test_ussd_api.py:
```diff
@@ -127,11 +141,66 @@ def test_golden_path_send_token(mocker, test_client,
     org = OrganisationFactory()
     sender = UserFactory(preferred_language="en", phone=make_kenyan_phone(phone()), first_name="Bob", last_name="Foo",
-        pin_hash=User.salt_hash_secret('0000'), default_organisation=org)
+                         pin_hash=User.salt_hash_secret('0000'), default_organisation=org)
```
why the extra indent? please either single indent or each arg on separate lines.

nolash commented on test/functional/app/api/test_ussd_api.py:
```diff
@@ -127,11 +141,66 @@ def test_golden_path_send_token(mocker, test_client,
+
+    # external auth password matching model definition of the same value.
+    # org.external_auth_username = 'admin_'+(org.name or '').lower().replace(' ', '_') /961ab9adc300_.py
+    external_auth_username = 'admin_' + (organisation.name or '').lower().replace(' ', '_')
```
Let's re-use the same code that creates it in the app as a function.

mango-habanero replied:
The code that creates in the app is a migration.
It is also really annoying, the discrepancy in what the name will be in the DB and what the config file is looking for

mango-habanero replied:
The code that creates it in the app is a migration, I don't think we can reuse it.

nolash replied:
Ok let's not spend time on this. If it's the same as in the migration that's fine then.

nolash commented on test/functional/app/api/test_ussd_api.py:
```diff
@@ -127,11 +141,66 @@ def test_golden_path_send_token(mocker, test_client,
+    resp = req("0000", test_client, user.phone, auth_username=external_auth_username)
+    assert "CON Enter your PIN again" in resp
+
+    resp = req("0000", test_client, user.phone, auth_username=external_auth_username)
```
Perhaps we should have a separate test for wrong pin, and retry with correct pins. This specific case made trouble at some point.

mango-habanero replied:
Noted, this will be added

nolash commented on test/functional/app/api/test_ussd_api.py:
```diff
@@ -2,6 +2,8 @@
 from faker.providers import phone_number
 from faker import Faker
 from functools import partial
+from contextlib import contextmanager
```
import not used?

mango-habanero replied:
Will clean up

nolash commented on test/functional/app/api/test_ussd_api.py:
```diff
@@ -2,6 +2,8 @@
 from faker import Faker
 from functools import partial
+from contextlib import contextmanager
+from flask import appcontext_pushed, g
```
`appcontext_pushed` not used?

mango-habanero replied:
Clean up as well

nolash commented on test/functional/app/api/test_ussd_api.py:
```diff
@@ -127,11 +141,66 @@ def test_golden_path_send_token(mocker, test_client,
+    assert bio.value.strip('"') == 'Unknown business'
+    gender = next(filter(lambda x: x.name == 'gender', user.custom_attributes), None)
+    assert gender.value.strip('"') == 'Unknown gender'
+    assert user.location == 'Unknown location'
```
Is it worth checking that the consecutive ussd session gives the usual menu?

nolash replied:
@PhilipWafula no comment on this?

nolash commented on test/functional/app/api/test_ussd_api.py:
```diff
@@ -127,11 +141,66 @@ def test_golden_path_send_token(mocker, test_client,
+    assert user.first_name == 'Unknown first name'
+    assert user.last_name == 'Unknown last name'
+    bio = next(filter(lambda x: x.name == 'bio', user.custom_attributes), None)
+    assert bio.value.strip('"') == 'Unknown business'
```
hmm suspicious that those values are returned encapsulated by  `"` ?

mango-habanero replied:
It would seem so. I'll confirm this though

mango-habanero committed:
Moves temporary user fixture to conftest.

mango-habanero committed:
Removes unused imports

mango-habanero committed:
Removes premature commit()

mango-habanero committed:
Improve self service wallet tests

mango-habanero committed:
Merge branch 'master' into philip/self-service-wallets-func-tests

mango-habanero committed:
Merges master

mango-habanero committed:
Merge branch 'philip/self-service-wallets-func-tests' of github.com:GrassrootsEconomics/SempoBlockchain into philip/self-service-wallets-func-tests

mango-habanero committed:
Merge branch 'master' into philip/self-service-wallets-func-tests

mango-habanero committed:
Cleans up tests.

nolash commented on test/unit/app/utils/test_utils_user.py:
```diff
@@ -18,7 +19,14 @@
 #     assert create_transfer_account_user.transfer_account is not None
 from helpers.factories import UserFactory, OrganisationFactory, TokenFactory, TransferAccountFactory, fake
-from server.utils.user import transfer_usages_for_user, send_onboarding_sms_messages, admin_reset_user_pin, proccess_create_or_modify_user_request
+from server.models.user import RegistrationMethodEnum
```
Oh please on one line, rather.

In general when we need so many symbols from one package that suggests we should have some sort of helper instead perhaps. We can look at that when and if we ever get around to cleaning up.

nolash commented on test/unit/app/utils/test_utils_user.py:
```diff
@@ -140,3 +148,29 @@ def test_send_welcome_sms(mocker, test_client, init_database, mock_sms_apis,
                         {'phone': '+61123456789', 'message': expected_terms},]
+
+def test_attach_transfer_account_to_user(test_client, init_database, create_organisation):
+    with db.session.no_autoflush:
```
why is this needed?

mango-habanero replied:
Tests that the transfer account is actually attached to a self service user once the post request to process the transfer account creation.
Feels like its implicitly tested in the functional tests for the self service flow now that you've brought it to my attentions.
Its components could largely just be incorporated to the test in the func tests.

nolash replied:
sorry I meant the autoflush - and we discucsed already :)

nolash commented on test/unit/app/utils/test_utils_user.py:
```diff
@@ -140,3 +148,29 @@ def test_send_welcome_sms(mocker, test_client, init_database, mock_sms_apis,
+        user_with_transfer_account = attach_transfer_account_to_user(user)
+
+        assert user_with_transfer_account.primary_blockchain_address is not None
+        bio = next(filter(lambda x: x.name == 'bio', user_with_transfer_account.custom_attributes), None)
```
are the custom attributes linked to the transfer account? if not, let's remove the checks that aren't in context.

nolash commented on test/functional/app/api/test_ussd_api.py:
```diff
@@ -2,81 +2,106 @@
+unregistered_user_phone = make_kenyan_phone(phone())
+
 
 @pytest.fixture(scope='module')
```
this fixture should probably also be somewhere else?

mango-habanero committed:
Merge branch 'master' into philip/self-service-wallets-func-tests

mango-habanero committed:
Clean up tests.

mango-habanero committed:
Refactors ussd tests to remove autoflush checks and improve tests..

mango-habanero committed:
Merge branch 'master' into philip/self-service-wallets-func-tests

nolash commented on test/functional/app/api/test_ussd_api.py:
```diff
@@ -236,3 +235,64 @@ def test_ussd_self_signup_wrong_pin_entry(test_client,
+    organisation.external_auth_password = config.EXTERNAL_AUTH_PASSWORD
+
+    # external auth password matching model definition of the same value.
+    # org.external_auth_username = 'admin_'+(org.name or '').lower().replace(' ', '_') /961ab9adc300_.py
```
please remove commented code

mango-habanero replied:
The comment isn't part of the code in anyway, it was there to provide context for why that `external_auth_name` was structured like that, in case we or someone else ever comes back to fix this, I feel the context would be helpful.
But then again it could be a less confusing comment so let me just make it clearer real quick.

nolash reviewed (commented):
@PhilipWafula thank you for your patience and for amending the code. Please see just one more small thing about commented code and then I will approve.

mango-habanero committed:
Cleans up test file.

- Adds cleaner comment for context on the creation of external_auth_username
- Removes unused definition of external_auth_username in test_reset_pin_flow

mango-habanero committed:
Merge branch 'philip/self-service-wallets-func-tests' of github.com:GrassrootsEconomics/SempoBlockchain into philip/self-service-wallets-func-tests
